translate-c is careless with @as() casting rhs of shift operators

(EDIT: i just noticed i'm switching between `_lrotr` and `_lrotl` in my examples here, but my point still stands.)

This C code:

https://github.com/ziglang/zig/blob/f83411b0b1b857c7f8679e3b90d2093ba60621d4/lib/libc/include/any-windows-any/psdk_inc/intrin-impl.h#L589-L593

gets translated as (formatted for some attempt at readability):

```zig
pub export fn _lrotr(arg___X: c_ulong, arg___C: c_int) c_ulong {
    var __X = arg___X;
    var __C = arg___C;
    return (
        (__X >> @as(@import("std").math.Log2Int(c_ulong), __C)) |
        (__X << @as(@import("std").math.Log2Int(c_ulong),
            ((
                @sizeOf(c_long) *%
                @bitCast(c_ulonglong, @as(c_longlong, @as(c_int, 8)))
            ) -%
                @bitCast(c_ulonglong, @as(c_longlong, __C))))));
}
```

The interesting part is that `__X << (something)` gets translated as `__X << @as(u5, (something))`, but that kind of casting will inevitably cause compile errors. It only happens to work when the rhs is a comptime-known int. Using `@as()` leads to this compile error:

```
Semantic Analysis [1029/1459] /home/josh/dev/www.legend-of-swarkland/deps/legend-of-swarkland/zig-cache/o/aAUHF1NHoPmktq9hU2jM5Sq5inPU_zBebEoZRAQZIyF9xUMTSfqK01IU90iZrSfr/cimport.zig:691:63: error: expected type 'u5', found 'c_int'
    return ((__X << @as(@import("std").math.Log2Int(c_ulong), __C)) | (__X >> @as(@import("std").math.Log2Int(c_ulong), ((@sizeOf(c_long) *% @bitCast(c_ulonglong, @as(c_longlong, @as(c_int, 8)))) -% @bitCast(c_ulonglong, @as(c_longlong, __C))))));
                                                              ^
/home/josh/dev/www.legend-of-swarkland/deps/legend-of-swarkland/zig-cache/o/aAUHF1NHoPmktq9hU2jM5Sq5inPU_zBebEoZRAQZIyF9xUMTSfqK01IU90iZrSfr/cimport.zig:691:63: note: unsigned 5-bit int cannot represent all possible signed 32-bit values
    return ((__X << @as(@import("std").math.Log2Int(c_ulong), __C)) | (__X >> @as(@import("std").math.Log2Int(c_ulong), ((@sizeOf(c_long) *% @bitCast(c_ulonglong, @as(c_longlong, @as(c_int, 8)))) -% @bitCast(c_ulonglong, @as(c_longlong, __C))))));
                                                              ^
```

I believe the correct conversion would be `@intCast(u5, (something))` instead.

--- a/src-self-hosted/translate_c.zig
+++ b/src-self-hosted/translate_c.zig
@@ -2778,13 +2778,13 @@ fn transCreateCompoundAssign(
             try transExpr(rp, scope, rhs, .used, .r_value);
 
         if (is_shift) {
-            const as_node = try transCreateNodeBuiltinFnCall(rp.c, "@as");
+            const cast_node = try transCreateNodeBuiltinFnCall(rp.c, "@intCast");
             const rhs_type = try qualTypeToLog2IntRef(rp, getExprQualType(rp.c, rhs), loc);
-            try as_node.params.push(rhs_type);
+            try cast_node.params.push(rhs_type);
             _ = try appendToken(rp.c, .Comma, ",");
-            try as_node.params.push(rhs_node);
-            as_node.rparen_token = try appendToken(rp.c, .RParen, ")");
-            rhs_node = &as_node.base;
+            try cast_node.params.push(rhs_node);
+            cast_node.rparen_token = try appendToken(rp.c, .RParen, ")");
+            rhs_node = &cast_node.base;
         }
         if (scope.id != .Condition)
             _ = try appendToken(rp.c, .Semicolon, ";");
@@ -2815,13 +2815,13 @@ fn transCreateCompoundAssign(
     const bin_token = try appendToken(rp.c, bin_tok_id, bin_bytes);
     var rhs_node = try transExpr(rp, scope, rhs, .used, .r_value);
     if (is_shift) {
-        const as_node = try transCreateNodeBuiltinFnCall(rp.c, "@as");
+        const cast_node = try transCreateNodeBuiltinFnCall(rp.c, "@intCast");
         const rhs_type = try qualTypeToLog2IntRef(rp, getExprQualType(rp.c, rhs), loc);
-        try as_node.params.push(rhs_type);
+        try cast_node.params.push(rhs_type);
         _ = try appendToken(rp.c, .Comma, ",");
-        try as_node.params.push(rhs_node);
-        as_node.rparen_token = try appendToken(rp.c, .RParen, ")");
-        rhs_node = &as_node.base;
+        try cast_node.params.push(rhs_node);
+        cast_node.rparen_token = try appendToken(rp.c, .RParen, ")");
+        rhs_node = &cast_node.base;
     }
     const rhs_bin = try transCreateNodeInfixOp(rp, scope, ref_node, bin_op, bin_token, rhs_node, .used, false);
 
@@ -3883,20 +3883,20 @@ fn transCreateNodeShiftOp(
     const lhs = try transExpr(rp, scope, lhs_expr, .used, .l_value);
     const op_token = try appendToken(rp.c, op_tok_id, bytes);
 
-    const as_node = try transCreateNodeBuiltinFnCall(rp.c, "@as");
+    const cast_node = try transCreateNodeBuiltinFnCall(rp.c, "@intCast");
     const rhs_type = try qualTypeToLog2IntRef(rp, ZigClangBinaryOperator_getType(stmt), rhs_location);
-    try as_node.params.push(rhs_type);
+    try cast_node.params.push(rhs_type);
     _ = try appendToken(rp.c, .Comma, ",");
     const rhs = try transExprCoercing(rp, scope, rhs_expr, .used, .r_value);
-    try as_node.params.push(rhs);
-    as_node.rparen_token = try appendToken(rp.c, .RParen, ")");
+    try cast_node.params.push(rhs);
+    cast_node.rparen_token = try appendToken(rp.c, .RParen, ")");
 
     const node = try rp.c.a().create(ast.Node.InfixOp);
     node.* = ast.Node.InfixOp{
         .op_token = op_token,
         .lhs = lhs,
         .op = op,
-        .rhs = &as_node.base,
+        .rhs = &cast_node.base,
     };
 
     return &node.base;

--- a/test/translate_c.zig
+++ b/test/translate_c.zig
@@ -814,7 +814,7 @@ pub fn addCases(cases: *tests.TranslateCContext) void {
         \\}
     , &[_][]const u8{
         \\pub export fn foo() c_int {
-        \\    return (@as(c_int, 1) << @as(@import("std").math.Log2Int(c_int), 2)) >> @as(@import("std").math.Log2Int(c_int), 1);
+        \\    return (@as(c_int, 1) << @intCast(@import("std").math.Log2Int(c_int), 2)) >> @intCast(@import("std").math.Log2Int(c_int), 1);
         \\}
     });
 
@@ -1980,7 +1980,7 @@ pub fn addCases(cases: *tests.TranslateCContext) void {
         \\    var a = arg_a;
         \\    var i: c_int = 0;
         \\    while (a > @bitCast(c_uint, @as(c_int, 0))) {
-        \\        a >>= @as(@import("std").math.Log2Int(c_int), 1);
+        \\        a >>= @intCast(@import("std").math.Log2Int(c_int), 1);
         \\    }
         \\    return i;
         \\}
@@ -2000,7 +2000,7 @@ pub fn addCases(cases: *tests.TranslateCContext) void {
         \\    var a = arg_a;
         \\    var i: c_int = 0;
         \\    while (a > @bitCast(c_uint, @as(c_int, 0))) {
-        \\        a >>= @as(@import("std").math.Log2Int(c_int), 1);
+        \\        a >>= @intCast(@import("std").math.Log2Int(c_int), 1);
         \\    }
         \\    return i;
         \\}
@@ -2051,14 +2051,14 @@ pub fn addCases(cases: *tests.TranslateCContext) void {
         \\        ref.* = ref.* ^ @as(c_int, 1);
         \\        break :blk ref.*;
         \\    });
-        \\    a >>= @as(@import("std").math.Log2Int(c_int), (blk: {
+        \\    a >>= @intCast(@import("std").math.Log2Int(c_int), (blk: {
         \\        const ref = &a;
-        \\        ref.* = ref.* >> @as(@import("std").math.Log2Int(c_int), @as(c_int, 1));
+        \\        ref.* = ref.* >> @intCast(@import("std").math.Log2Int(c_int), @as(c_int, 1));
         \\        break :blk ref.*;
         \\    }));
-        \\    a <<= @as(@import("std").math.Log2Int(c_int), (blk: {
+        \\    a <<= @intCast(@import("std").math.Log2Int(c_int), (blk: {
         \\        const ref = &a;
-        \\        ref.* = ref.* << @as(@import("std").math.Log2Int(c_int), @as(c_int, 1));
+        \\        ref.* = ref.* << @intCast(@import("std").math.Log2Int(c_int), @as(c_int, 1));
         \\        break :blk ref.*;
         \\    }));
         \\}
@@ -2109,14 +2109,14 @@ pub fn addCases(cases: *tests.TranslateCContext) void {
         \\        ref.* = ref.* ^ @bitCast(c_uint, @as(c_int, 1));
         \\        break :blk ref.*;
         \\    });
-        \\    a >>= @as(@import("std").math.Log2Int(c_uint), (blk: {
+        \\    a >>= @intCast(@import("std").math.Log2Int(c_uint), (blk: {
         \\        const ref = &a;
-        \\        ref.* = ref.* >> @as(@import("std").math.Log2Int(c_int), @as(c_int, 1));
+        \\        ref.* = ref.* >> @intCast(@import("std").math.Log2Int(c_int), @as(c_int, 1));
         \\        break :blk ref.*;
         \\    }));
-        \\    a <<= @as(@import("std").math.Log2Int(c_uint), (blk: {
+        \\    a <<= @intCast(@import("std").math.Log2Int(c_uint), (blk: {
         \\        const ref = &a;
-        \\        ref.* = ref.* << @as(@import("std").math.Log2Int(c_int), @as(c_int, 1));
+        \\        ref.* = ref.* << @intCast(@import("std").math.Log2Int(c_int), @as(c_int, 1));
         \\        break :blk ref.*;
         \\    }));
         \\}

